Realm/Tests/AsyncTests.m

```diff
@@ -0,0 +1,808 @@
+    id token = [[array.intArray objectsWhere:@"intCol > 0"] deliverOn:queue block:^(RLMResults *results, NSError *e) {
+        XCTAssertNil(e);
+        XCTAssertNotNil(results);
+        XCTAssertEqual((int)results.count, expected);
+        for (int i = 0; i < expected; ++i) {
+            XCTAssertEqual([results[i] intCol], i + 1);
+        }
+        ++expected;
+        dispatch_semaphore_signal(sema);
+    }];
+    dispatch_semaphore_wait(sema, DISPATCH_TIME_FOREVER);
+
+    for (int i = 0; i < 3; ++i) {
+        // Create two objects, one in the list and one not, to verify that the
+        // LinkList is actually be used
+        [realm beginWriteTransaction];
+        [IntObject createInRealm:realm withValue:@[@(i + 1)]];
+        [array.intArray addObject:[IntObject createInRealm:realm withValue:@[@(i + 1)]]];
+        [realm commitWriteTransaction];
+        dispatch_semaphore_wait(sema, DISPATCH_TIME_FOREVER);
+    }
+    dispatch_sync(queue, ^{});
+    [token stop];
+}
+
+- (RLMNotificationToken *)subscribeAndWaitForInitial:(RLMResults *)query block:(void (^)(RLMResults *))block {
+    __block XCTestExpectation *exp = [self expectationWithDescription:@"wait for initial results"];
+    id token = [query deliverOnMainThread:^(RLMResults *results, NSError *e) {
+        XCTAssertNotNil(results);
+        XCTAssertNil(e);
+        if (exp) {
+            [exp fulfill];
+            exp = nil;
+        }
+        else {
+            block(results);
+        }
+    }];
+    [self waitForExpectationsWithTimeout:2.0 handler:nil];
+    return token;
+}
+
+- (void)testManualRefreshUsesAsyncResultsWhenPossible {
+    __block bool called = false;
+    id token = [self subscribeAndWaitForInitial:IntObject.allObjects block:^(RLMResults *results) {
+        called = true;
+    }];
+
+    RLMRealm *realm = [RLMRealm defaultRealm];
+    realm.autorefresh = NO;
+
+    [self waitForNotification:RLMRealmRefreshRequiredNotification realm:realm block:^{
+        [self dispatchAsync:^{
+            [RLMRealm.defaultRealm transactionWithBlock:^{
+                [IntObject createInDefaultRealmWithValue:@[@0]];
+            }];
+        }];
+    }];
+
+    XCTAssertFalse(called);
+    [realm refresh];
+    XCTAssertTrue(called);
+
+    [token stop];
+}
+
+- (void)testModifyingUnrelatedTableDoesNotTriggerResend {
+    __block XCTestExpectation *expectation = [self expectationWithDescription:@""];
+    id token = [[IntObject allObjects] deliverOnMainThread:^(RLMResults *results, NSError *e) {
+        // will throw if called a second time
+        [expectation fulfill];
+    }];
+    [self waitForExpectationsWithTimeout:2.0 handler:nil];
+
+    [self waitForNotification:RLMRealmDidChangeNotification realm:RLMRealm.defaultRealm block:^{
+        RLMRealm *realm = [RLMRealm defaultRealm];
+        [realm transactionWithBlock:^{
+            [StringObject createInDefaultRealmWithValue:@[@""]];
+        }];
+    }];
+    [token stop];
```

Comment:
There's nothing to wait for. This test relies on the fact that Realm notifications are sent after new results.